### DISCUSSION TITLE
아래 코드를 실행했을 때 결과가 어떻게 될까요?

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#106] [모듈]
- 아래 코드를 실행했을 때 결과가 어떻게 될까요?
  ```jsx
  // math.js
  var x = 5;
  
  export function addFromFive(a, b) {
    return this.x + a + b;
  }
 
  // 파일 분리 -------------

  // main.js
  import { addFromFive } from './math.js';
  
  var x = 6;
  
  console.log(addFromFive(2, 3));
  ```

**Comment by publdaze** [comment]
> * 아래 코드를 실행했을 때 결과가 어떻게 될까요?
>   ```js
>   // math.js
>   var x = 5;
>   
>   export function addFromFive(a, b) {
>     return this.x + a + b;
>   }
>   
>   // 파일 분리 -------------
>   
>   // main.js
>   import { addFromFive } from './math.js';
>   
>   var x = 6;
>   
>   console.log(addFromFive(2, 3));
>   ```

- main이 평가되고, 실행 시작. 1번라인인 import문 실행
- math.js로 넘어가 코드 평가와 실행이 이루어짐
- 이때, x는 모듈 독자적인 스코프의 지역 변수로 5라는 값이 할당 됨
- 이후 다시 main으로 돌아와서
- 1번 라인 이후부터 이어서 실행 x에 값이 할당되는데 여기서 x는 main 독자적인 스코프의 식별자이므로 math의 x와 공유되지 않고 6이 할당됨
- 이후 addFromFive가 호출되면 해당 함수의 this가 바인딩되는데, ES6 모듈에서는 항상 strict 모드이기 때문에 this는 undefined가 됨
- 따라서 this.x에서 (undefined).x 로 평가되고, undefined에는 x라는 프로퍼티가 존재하지 않음으로 type에러 발생

**Comment by pipisebastian** [comment]
> * 아래 코드를 실행했을 때 결과가 어떻게 될까요?
>   ```js
>   // math.js
>   var x = 5;
>   
>   export function addFromFive(a, b) {
>     return this.x + a + b;
>   }
>   
>   // 파일 분리 -------------
>   
>   // main.js
>   import { addFromFive } from './math.js';
>   
>   var x = 6;
>   
>   console.log(addFromFive(2, 3));
>   ```

- main의 `addFromFive` 는 모듈로부터 import 한 함수
    - `addFromFive`
        - 매개변수로 받은 `a`, `b` 존재
        - `this.x`를 참조해야함
            - 그런데, `addFromFive` 는 일반함수로써 호출됨
            - ES6 모듈 에서는 strict mode → 일반 함수로써 호출했을때, `this`에는 undefined가 바인딩됨
            - 그렇기 때문에, `this`는 undefined 이고, this.x 도 존재하지 않음 ⇒ `TypeError`
